x-pack/libbeat/management/unit.go

```diff
@@ -0,0 +1,400 @@
+		streamIDs:    streamIDs,
+		streamStates: streamStates,
+	}
+}
+
+// RegisterAction registers action handler for this unit.
+func (u *agentUnit) RegisterAction(action client.Action) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return
+	}
+
+	u.clientUnit.RegisterAction(action)
+}
+
+// UnregisterAction unregisters action handler with the client.
+func (u *agentUnit) UnregisterAction(action client.Action) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return
+	}
+
+	u.clientUnit.UnregisterAction(action)
+}
+
+func (u *agentUnit) Expected() client.Expected {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return client.Expected{}
+	}
+
+	return u.clientUnit.Expected()
+}
+
+func (u *agentUnit) ID() string {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return ""
+	}
+
+	return u.clientUnit.ID()
+}
+
+// calcState calculates the current state of the unit.
+func (u *agentUnit) calcState() (status.Status, string) {
+	// for type output return the unit state directly as it has no streams
+	if u.clientUnit.Type() == client.UnitTypeOutput {
+		return u.input.state, u.input.msg
+	}
+
+	// if input state is not running return the input state
+	if u.input.state != status.Running {
+		return u.input.state, u.input.msg
+	}
+
+	// input state is marked as running, check the stream states
+	currStatus := status.Running
+	countFailed := 0
+	countDegraded := 0
+	for _, streamState := range u.streamStates {
+		switch streamState.state {
+		case status.Degraded:
+			countDegraded++
+			if currStatus == status.Running {
+				currStatus = status.Degraded
+			}
+		case status.Failed:
+			countFailed++
+			currStatus = status.Failed
+		}
+	}
+
+	if currStatus == status.Running {
+		return currStatus, "Healthy"
+	}
+
+	return currStatus, fmt.Sprintf("Out of %d streams, %d are failed, %d are degraded",
+		len(u.streamStates), countFailed, countDegraded)
+}
+
+// Type of the unit.
+func (u *agentUnit) Type() client.UnitType {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return client.UnitTypeInput
+	}
+
+	return u.clientUnit.Type()
+}
+
+// UpdateState updates the state for the unit.
+func (u *agentUnit) UpdateState(state status.Status, msg string, payload map[string]interface{}) error {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return nil
+	}
+
+	if u.input.state == state && u.input.msg == msg {
+		return nil
+	}
+
+	u.input = unitState{
+		state: state,
+		msg:   msg,
+	}
+
+	state, msg = u.calcState()
+
+	if u.clientUnit.Type() == client.UnitTypeOutput || len(u.streamStates) == 0 {
+		return u.clientUnit.UpdateState(getUnitState(state), msg, payload)
+	}
+
+	streamsPayload := make(map[string]interface{}, len(u.streamStates))
+
+	for streamID, streamState := range u.streamStates {
+		streamsPayload[streamID] = map[string]interface{}{
+			"status": getUnitState(streamState.state).String(),
+			"error":  streamState.msg,
```

Comment:
I suggested this in a top level comment as well, I would rather see the message be the error from the first failed stream, rather than a summary. Otherwise we are going to get support cases asking to explain what the error actually is, users don't know to look at the raw agent JSON document in Fleet.

Reply:
I will reply to the top level comment